sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/StreamingSessionWindowStateManager.scala

```diff
@@ -0,0 +1,370 @@
+  @transient private lazy val keyRowGenerator = UnsafeProjection.create(
+    keyWithoutSessionExpressions, valueAttributes)
+  @transient private lazy val stateKeyRowGenerator = UnsafeProjection.create(stateKeyExprs,
+    keyWithoutSessionExpressions)
+
+  @transient private lazy val helper = new StreamingSessionWindowHelper(
+    sessionExpression, valueAttributes)
+
+  override def getStateKeySchema: StructType = stateKeyStructType
+
+  override def getStateValueSchema: StructType = valueAttributes.toStructType
+
+  override def getNumColsForPrefixKey: Int = keyWithoutSessionExpressions.length
+
+  override def extractKeyWithoutSession(value: UnsafeRow): UnsafeRow = {
+    keyWithoutSessionRowGenerator(value)
+  }
+
+  override def newOrModified(store: ReadStateStore, value: UnsafeRow): Boolean = {
+    val sessionStart = helper.extractTimePair(value)._1
+    val stateKey = getStateKey(getKey(value), sessionStart)
+    val stateRow = store.get(stateKey)
+    stateRow == null || !stateRow.equals(value)
+  }
+
+  override def getSessions(store: ReadStateStore, key: UnsafeRow): Iterator[UnsafeRow] =
+    getSessionsWithKeys(store, key).map(_.value)
+
+  private def getSessionsWithKeys(
+      store: ReadStateStore,
+      key: UnsafeRow): Iterator[UnsafeRowPair] = {
+    store.prefixScan(key)
+  }
+
+  override def updateSessions(
+      store: StateStore,
+      key: UnsafeRow,
+      sessions: Seq[UnsafeRow]): Long = {
+    replaceModifyWindow(key, sessions, store)
+  }
+
+  override def commit(store: StateStore): Long = store.commit()
+
+  override def iterator(store: ReadStateStore): Iterator[UnsafeRow] =
+    iteratorWithKeys(store).map(_.value)
+
+  private def iteratorWithKeys(store: ReadStateStore): Iterator[UnsafeRowPair] = {
+    store.iterator()
+  }
+
+  override def removeByValueCondition(
+      store: StateStore,
+      removalCondition: UnsafeRow => Boolean): Iterator[UnsafeRow] = {
+    new NextIterator[UnsafeRow] {
+      private val rangeIter = iteratorWithKeys(store)
+
+      override protected def getNext(): UnsafeRow = {
+        var removedValueRow: UnsafeRow = null
+        while(rangeIter.hasNext && removedValueRow == null) {
+          val rowPair = rangeIter.next()
+          if (removalCondition(rowPair.value)) {
+            store.remove(rowPair.key)
+            removedValueRow = rowPair.value
+          }
+        }
+
+        if (removedValueRow == null) {
+          finished = true
+          null
+        } else {
+          removedValueRow
+        }
+      }
+
+      override protected def close(): Unit = {}
+    }
+  }
+
+  private def getKey(value: UnsafeRow): UnsafeRow = keyRowGenerator(value)
+
+  private def getStateKey(key: UnsafeRow, sessionStart: Long): UnsafeRow = {
+    val stateKey = stateKeyRowGenerator(key)
+    stateKey.setLong(indexOrdinalInSessionStart, sessionStart)
+    stateKey.copy()
+  }
+
+  /**
+   * add to store, only update the changed window elements
```

Comment:
Could you rephrase the method doc? Maybe "Updates windows from the values overlapped with existing values in the store and adds new windows from the values."?